Add work with database.

lazyabrikos opened the pull request:
Add classes which working with  db.

lazyabrikos committed:
Add work with database.
Add some Unit-tests.

lazyabrikos committed:
code style

lazyabrikos committed:
code style

lazyabrikos committed:
code style

lazyabrikos committed:
drop tables

TyutterinYakov commented on pom.xml:
```diff
@@ -49,6 +49,21 @@
 		</dependency>
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-data-jpa</artifactId>
```
Действительно ли эта зависимость требуется?) JPA это еще более высокая абстракция над JDBC и будет изучена позже)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/GenreController.java:
```diff
@@ -0,0 +1,39 @@
+@RestController
+@RequestMapping("/genres")
+public class GenreController {
+    private static final Logger logger = LoggerFactory.getLogger(GenreController.class);
```
Можно было бы использовать аннотацию из lombok ->> [@Slf4j](https://www.slf4j.org/manual.html)
Аналогично в других местах

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/MpaController.java:
```diff
@@ -0,0 +1,40 @@
+@RestController
+@RequestMapping("/mpa")
+public class MpaController {
+    private static final Logger logger = LoggerFactory.getLogger(MpaController.class);
```
Можно было бы использовать аннотацию из lombok ->> [@Slf4j](https://www.slf4j.org/manual.html)
Аналогично в других местах

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/GenreController.java:
```diff
@@ -0,0 +1,39 @@
+    private static final Logger logger = LoggerFactory.getLogger(GenreController.class);
+    private final GenreService genreService;
+
+    public GenreController(GenreService genreService) {
```
Так же можно воспользоваться аннотацией `@RequiredArgsConstructor` из lombok, благодаря которой автоматически будет сгенерирован конструктор для финальных полей)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/MpaController.java:
```diff
@@ -0,0 +1,40 @@
+    private static final Logger logger = LoggerFactory.getLogger(MpaController.class);
+    private final MpaService mpaService;
+
+    public MpaController(MpaService mpaService) {
```
Так же можно воспользоваться аннотацией `@RequiredArgsConstructor` из lombok, благодаря которой автоматически будет сгенерирован конструктор для финальных полей)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/MpaController.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    @GetMapping
+    @ResponseStatus(HttpStatus.OK)
```
Статус 200 OK является кодом ответа по умолчанию) Можно было бы не добавлять в таком случае `@ResponseStatus`
Аналогично в других местах

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/GenreController.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    @GetMapping
+    @ResponseStatus(HttpStatus.OK)
```
Статус 200 OK является кодом ответа по умолчанию) Можно было бы не добавлять в таком случае `@ResponseStatus`
Аналогично в других местах

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/errors/ErrorHandler.java:
В класс следует добавить логирование) Для этого достаточно использовать аннотацию из lombok ->> `@Slf4j`. Далее в каждый наш хендлер следует добавить логирование сообщения об ошибке)
Также хорошо было бы передать в лог сообщение об ошибке, а можно и весь трейс, чтобы в случае ошибки было проще разобраться, что именно произошло)
Кстати, при логгировании более эффективно использовать не конкатенацию, а заполнитель - placeholder `{}`
Например:
`log.debug("Получен статус 404 Not found {}", ex.getMessage(), ex);`

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/errors/ErrorHandler.java:
```diff
@@ -55,4 +56,10 @@ public ErrorResponse handleServerErrors(final Exception e) {
 
+    @ExceptionHandler
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public ErrorResponse handleIncorrectParameter(final IncorrectFieldException e) {
```
Так же можно было бы не создавать для каждого исключения, обрабатываемого как `BAD_REQUEST`, отдельные хендлеры, а собрать их в одну группу в `@ExceptionHandler({IncorrectFieldException.class, NegativeValueException.class...})`
А принимаемым аргументом метода-обработчика будет общий класс-родитель - `Exception`

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -33,13 +37,20 @@ public class Film {
-    public Film(String name, String description, LocalDate releaseDate, int duration) {
+    private Mpa mpa;
+
+    private Collection<Genre> genres = new ArrayList<>();
```
Для хранения жанров следует использовать `LinkedHashSet`, так как нам важно поддерживать уникальность и сохранять порядок

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -33,13 +37,20 @@ public class Film {
+    private Long duration;
 
-    public Film(String name, String description, LocalDate releaseDate, int duration) {
+    private Mpa mpa;
```
Так как фильмов без рейтинга не должно быть в нашей системе - хорошо было бы навесить над полем `@NotNull`

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
Над полем с описанием хорошо было бы также навесить `@NotBlank`, чтобы не допустить `null`, пустую или состоящую лишь из пробелов строку)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
Хорошо было бы ограничить размеры текстовых полей на основе тех ограничений, которые установлены на уровне таблицы в БД, чтобы заранее откинуть некорректные запросы) Для ограничения количества символов можно использовать `@Size`

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
Список с лайками более не требуется, так как хранение теперь осуществляется в БД

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -6,16 +6,19 @@
     private Long id;
+
     @JsonIgnore
     private Set<Long> friends = new HashSet<>();
```
Список с идентификаторами друзей пользователя более не требуется, так как хранение теперь осуществляется в БД

TyutterinYakov commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,68 @@
+(
+    id int generated by default as identity primary key,
+    name varchar(255),
+    login varchar(255) not null,
```
Можно было бы ввести ограничение на уникальность логина и почты на уровне таблицы в БД)
```
create unique index if not exists USER_EMAIL_UINDEX on USERS (email);
create unique index if not exists USER_LOGIN_UINDEX on USERS (login);
```

TyutterinYakov commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,68 @@
+(
+    user_id int,
+    friend_id int,
+    relation_accept varchar,
```
Хранение вычисляемых полей на уровне БД - не является хорошей практикой. Это пример денормализации, к которой прибегают в целях оптимизации, когда других вариантов не остается)
Поле лучше исключить

TyutterinYakov commented on src/main/resources/data.sql:
```diff
@@ -0,0 +1,12 @@
+insert into mpa (name) values ('G');
```
Можно было бы использовать синтаксис:
`INSERT INTO ... VALUES (...), (...), ...`
Чтобы произвести вставку за два запроса)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -1,20 +1,23 @@
+    private final FriendsStorage friendsStorage;
 
-    public UserService(UserStorage userStorage) {
+    public UserService(@Qualifier("userDbStorage") UserStorage userStorage, FriendsStorage friendsStorage) {
```
Действительно ли требуется использовать `@Qualifier`
Интерфейс `UserStorage` имеет лишь одну реализацию) InMemory-реализация более не требуется

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,50 +1,128 @@
 
-
-    public FilmService(FilmStorage filmStorage, UserService userService) {
+    public FilmService(@Qualifier("filmDbStorage") FilmStorage filmStorage,
```
Действительно ли требуется использовать `@Qualifier`
Интерфейс `FilmStorage` имеет лишь одну реализацию) InMemory-реализация более не требуется

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,50 +1,128 @@
-    public FilmService(FilmStorage filmStorage, UserService userService) {
+    public FilmService(@Qualifier("filmDbStorage") FilmStorage filmStorage,
+                       UserService userService, MpaService mpaService,
+                       GenreDbStorage genreDbStorage, LikeStorage likeDbStorage, GenreService genreService) {
```
Лучше не использовать при именовании `Db`, так как мы работаем лишь с предоставляемым API интерфейса и реализация может быть заменена, а название будет вводить в заблуждение)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,50 +1,128 @@
+        }
+
+        film = filmStorage.create(film);
+        System.out.println(genresIds);
```
Для логгирования следует использовать логгер

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,50 +1,128 @@
+        List<Genre> genres;
+        try {
+            genres = genresIds.stream()
+                    .map(genreService::getById)
```
При текущей реализации мы плодим n-лишние запросы, так как производим обращение к БД в цикле) Следует произвести получение с использованием SQL условия IN

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/genres/GenreDbStorage.java:
```diff
@@ -0,0 +1,59 @@
+
+    @Override
+    public void insertFilmAndGenres(long filmId, Set<Long> genresId) {
+        for (Long id : genresId) {
```
В данном случае вместо цикла следует использовать `jdbcTemplate.batchUpdate(...`, эта логика будет находиться в репозитории жанров)
Таким образом мы не будем плодить n-лишние запросы к БД, а сначала соберем все в один пакет и отправим одним запросом)
Либо же можно использовать `StringBuilder`, чтобы сначала сформировать запрос, а затем отправив его за раз)
https://mkyong.com/spring/spring-jdbctemplate-batchupdate-example/

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/GenreService.java:
```diff
@@ -0,0 +1,37 @@
+
+@Service
+public class GenreService {
+    private final GenreStorage genreDbStorage;
```
Также при именовании лучше не указывать `Db`

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,50 +1,128 @@
+        List<Genre> genres;
+        try {
+            genres = genresIds.stream()
+                    .map(genreService::getById)
```
Здесь также мы плодим n-лишние запросы)
О том, как использовать SQL оператор условия `IN` для получение по коллекции идентификаторов - можно посмотреть тут https://www.baeldung.com/spring-jdbctemplate-in-list

TyutterinYakov reviewed (commented):
Добрый день! В целом - хорошо) Отметил то, что следует исправить и то, что можно сделать лучше) Хорошего дня! Если возникнут вопросы по ревью - пиши в пачку, буду рад помочь🌅

lazyabrikos committed:
Code refactor.
Remake friendship table, remove relation_accept field;
Rework Genre and Film Insert to BatchUpdate;

lazyabrikos committed:
Code style refactor

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -33,13 +35,21 @@ public class Film {
 
     @Positive(message = "Duration must be positive.")
-    private int duration;
+    private Long duration;
```
Если меняем тип на ссылочный, следует также навесить `@NotNull`, так как `@Positive` допускает `null` за корректное значение

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/genres/GenreDbStorage.java:
```diff
@@ -0,0 +1,86 @@
+
+    @Override
+    public void insertFilmAndGenres(long filmId, Set<Long> genresId) {
+        for (Long id : genresId) {
```
Эта логика была заменена запись ниже) Следует эту исключить

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/genres/GenreStorage.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.Set;
+
+public interface GenreStorage {
+    Genre getGenreById(long id);
```
(Предложение) Хорошо было бы использовать в методах получения по идентификатору тип `Optional` https://habr.com/ru/post/346782/ Гарантирую, ты не пожалеешь, а заодно в будущем тебе будет проще, так как он активно используется при работе со Spring Data JPA

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/GenreService.java:
```diff
@@ -0,0 +1,40 @@
+    private final GenreStorage genreStorage;
+
+    public Genre getById(Long id) {
+        return genreStorage.getGenreById(id);
```
Если будешь использовать тип Optional, то здесь как раз будет проверка, можно будет использовать простую лямбду `.orElseThrow(() -> new NotFoundExcetion(...));` То есть, если объект не будет найден, будет выброшено исключение, а если он есть, мы просто его получим, как будто написали .`get()`
Это для метода получения по идентификатору, к которому, кстати, можно потом обращаться, чтобы не дублировать строку с выбросом исключения)

TyutterinYakov reviewed (approved):
Добрый день! Отметил оставшиеся недочеты, но думаю, что ты справишься с ними вне ревью) Успехов в обучении и прекрасного дня!🌅🎇

lazyabrikos committed:
Refactor storages, return Optional<Object>

lazyabrikos committed:
Code style refactor

lazyabrikos committed:
Code style refactor
